content/blog/posts/2022-08-04-add-desktop-notifications.md

```diff
@@ -0,0 +1,241 @@
+---
+authorId: chuckcarpenter
+categories:
+  - notifications
+  - javascript
+date: ''
+slug: add-desktop-notifications
+title: Adding desktop notifications to your app/site
+---
+
+There are times when your application needs to inform users to act on something more
+timely than a normal notification (often known as toast notifications) may allow.
+If you need a user to do something and can't wait until they return to the application for them to notice, or just want to engage them when something new is available, desktop notifications might be a good fit for your app.
+
+Desktop notifications are supported in all modern browsers (sorry Internet Explorer, you're not invited). However, the user will have to opt in and they will also have to allow notifications in their OS settings. One they have opted in, you can notify them on the system level rather than just within your web browser page/tab.
+
+## What we'll work on today
+
+Let's create a basic notification that checks for and then gets the user's permission. Then we can show a notification about our cool new blog posts. We can then add some actions to allow the user to opt it future notifications as to when new posts are published. There's more to it, but we'll get to the nuances shortly.
+
+For this I'm going to create a singleton pattern file since I need it to load and run when loaded. I find this pattern lends itself well to our needs and is the most readable.
+
+```js
+class NotificationService {
+  constructor() {
+    // whatever you want once instance is created
+  }
+
+  async _isShowNotificationPermitted() {
+    // ...
+  }
+
+  async _showNotification() {
+    // ...
+  }
+}
+
+new NotificationService();
+```
+
+Using this javascript class I am adding a couple methods to handle checking the permission settings, requesting permissions, and then showing a notification if granted. Let's start with checking for and asking for permission to show notifications from our site.  
+
+```js
+async _isShowNotificationPermitted() {
+  if ('Notification' in window) {
+    let { permission } = Notification;
+    if (permission !== 'granted') {
+      permission =  await Notification.requestPermission();
+    }
+    if (permission !== 'denied') {
```

Comment:
This is probably not needed because I'm not sure how it would come back as `default` here, but a belt and suspenders approach might also cover `default` here so all possible values are handled.

```suggestion
    if (permission !== 'denied' || permission !== 'default') {
```